### PR TITLE
Reproduce invalid deprecated_library_name

### PR DESCRIPTION
**Pull request by rgrinberg**
The following redirect isn't allowed:

```dune
(deprecated_library_name (old_public_name a) (new_public_name a))
```

But we don't validate this use case.

I'd like to fix this, but now I'm thinking that it might be possible to
introduce longer redirect cycles. As far as I know, we only look detect cycles
when instantiating libraries, so a cycle involving redirects should be possible.

@nojb @jeremiedimino  does that sound right?

### CODE REVIEW

**Comment by rgrinberg** [comment]
@nojb I've added another test case for what I think is some weirdness from the `deprecated_library_name` stanza. `old_public_name` allows us to refer to private names as well as public names which seems wrong.

**Comment by nojb** [comment]
I don't remember the details (and the code has been refactored since I last touched it), but indeed the examples look fishy.

**Comment by rgrinberg** [comment]
@nojb this is now ready as well

**Comment by jeremiedimino** [comment]
@rgrinberg I remember that you mentioned that you wanted to extend this test to detect all cycles involving deprecated library names. However, this PR only detect cycles of length 1, is this PR still relevant as it?

**Comment by rgrinberg** [comment]
Yeah, that was my goal but in the end I shelved that work because I couldn't reproduce a longer cycle in a test. I think this PR is still good as it's replacing a crash with a decent error message in an edge case.

**Comment by jeremiedimino** [comment]
Ack

File: test/blackbox-tests/test-cases/deprecated-library-name/name-clash.t [hunk 1]
```diff
@@ -0,0 +1,20 @@
+The old_public_name field is named in a confusing matter. As we can see in this
```
Comment (jeremiedimino) [inline test/blackbox-tests/test-cases/deprecated-library-name/name-clash.t:1]:
```suggestion
The old_public_name field is named in a confusing manner. As we can see in this
```

File: test/blackbox-tests/test-cases/deprecated-library-name/name-clash.t [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+test, the name may refer to private names in the current scope as well as public
+names.
```
Comment (jeremiedimino) [inline test/blackbox-tests/test-cases/deprecated-library-name/name-clash.t:3]: The explanation doesn't seem correct to me. `(old_public_name b)` doesn't refer to an existing name, it defines a new name `b` as a redirection to another one. You'd have the same error message if you replaced `(deprecated_library_name ...)` by `(library (public_name b))`.

In this example when Dune tries to understand what `b` means it has two choices:

- `b` can be interpreted as one of the deprecated name of library `a`
- `b` can be interpreted as the private library `b`

both choices are valid, so Dune can't make a clear decision.

File: test/blackbox-tests/test-cases/deprecated-library-name/redirect-cycle.t [hunk 3]
```diff
@@ -0,0 +1,17 @@
+Creating an invalid (deprecated_libary_name ..) is an error
```
Comment (jeremiedimino) [inline test/blackbox-tests/test-cases/deprecated-library-name/redirect-cycle.t:1]: What is the difference between this test and `invalid-redirect.t`?

**Review by jeremiedimino — CHANGES_REQUESTED** [review]

File: test/blackbox-tests/test-cases/deprecated-library-name/redirect-cycle.t (hunk 3 above)
Comment (rgrinberg) [reply test/blackbox-tests/test-cases/deprecated-library-name/redirect-cycle.t:1]: It was a botched renamed. Good catch.

File: test/blackbox-tests/test-cases/deprecated-library-name/name-clash.t (hunk 2 above)
Comment (rgrinberg) [reply test/blackbox-tests/test-cases/deprecated-library-name/name-clash.t]: I suppose this test doesn't really clarify anything then. I'm removing it.

**Review by jeremiedimino — APPROVED** [review]
Looks good

**Pull request merged**